docs/retries.md

```diff
@@ -0,0 +1,338 @@
+# HTTP Request Retries
+
+Pydantic AI provides retry functionality for HTTP requests made by model providers through custom HTTP transports.
+This is particularly useful for handling transient failures like rate limits, network timeouts, or temporary server errors.
+
+## Overview
+
+The retry functionality is built on top of the [tenacity](https://github.com/jd/tenacity) library and integrates
+seamlessly with httpx clients. You can configure retry behavior for any provider that accepts a custom HTTP client.
+
+## Installation
+
+To use the retry transports, you need to install `tenacity`, which you can do via the `retries` dependency group:
+
+```bash
+pip/uv-add 'pydantic-ai-slim[retries]'
+```
+
+## Usage Example
+
+Here's an example of adding retry functionality with smart retry handling:
+
+```python {title="smart_retry_example.py"}
+from httpx import AsyncClient, HTTPStatusError
+from tenacity import (
+    AsyncRetrying,
+    stop_after_attempt,
+    wait_exponential,
+    retry_if_exception_type
+)
+from pydantic_ai import Agent
+from pydantic_ai.models.openai import OpenAIModel
+from pydantic_ai.retries import AsyncTenacityTransport, wait_retry_after
+from pydantic_ai.providers.openai import OpenAIProvider
+
+def create_retrying_client():
+    """Create a client with smart retry handling for multiple error types."""
+
+    def should_retry_status(response):
+        """Raise exceptions for retryable HTTP status codes."""
+        if response.status_code in (429, 502, 503, 504):
+            response.raise_for_status()  # This will raise HTTPStatusError
```

Comment:
i've been trying to follow this approach, but for some reason, when the code reaches this line it explodes because the request is not set on the response object.
i wonder if it's related to this new change